File: shinkai-message-wasm/src/shinkai_message/shinkai_message.rs
use serde::{Deserialize, Serialize};
use crate::shinkai_utils::encryption::EncryptionMethod;
use super::shinkai_message_schemas::MessageSchemaType;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ShinkaiMessage {
    pub body: MessageBody,
    pub external_metadata: ExternalMetadata,
    pub encryption: EncryptionMethod,
    pub version: ShinkaiVersion
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ShinkaiBody {
    pub message_data: MessageData,
    pub internal_metadata: InternalMetadata,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InternalMetadata {
    pub sender_subidentity: String,
    pub recipient_subidentity: String,
    pub inbox: String,
    pub signature: String,
    pub encryption: EncryptionMethod,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExternalMetadata {
    pub sender: String,
    pub recipient: String,
    pub scheduled_time: String,
    pub signature: String,
    pub other: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EncryptedShinkaiBody {
    pub content: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct EncryptedShinkaiData {
    pub content: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ShinkaiData {
    pub message_raw_content: String,
    pub message_content_schema: MessageSchemaType,
}

#[derive(Debug, Clone)]
pub enum MessageBody {
    Encrypted(EncryptedShinkaiBody),
    Unencrypted(ShinkaiBody),
}

#[derive(Debug, Clone, PartialEq)]
pub enum MessageData {
    Encrypted(EncryptedShinkaiData),
    Unencrypted(ShinkaiData),
}

#[derive(Debug, Clone, PartialEq)]
pub enum ShinkaiVersion {
    V1_0,
    Unsupported,
}

File: shinkai-app/src/models/ShinkaiMessage.ts
export enum EncryptionMethod {
  DiffieHellmanChaChaPoly1305 = "DiffieHellmanChaChaPoly1305",
  None = "None",
}

export enum MessageSchemaType {
  JobCreationSchema = "JobCreationSchema",
  JobMessageSchema = "JobMessageSchema",
  PreMessageSchema = "PreMessageSchema",
  TextContent = "TextContent",
  Empty = "",
}

export interface JobScope {
  buckets: string[];
  documents: string[];
}

export interface JobCreation {
  scope: JobScope;
}

export interface JobMessage {
  job_id: string;
  content: string;
}

export interface JobToolCall {
  tool_id: string;
  inputs: Record<string, string>;
}

export enum JobRecipient {
  SelfNode = "SelfNode",
  User = "User",
  ExternalIdentity = "ExternalIdentity",
}

export interface JobPreMessage {
  tool_calls: JobToolCall[];
  content: string;
  recipient: JobRecipient;
}

export interface InternalMetadata {
  sender_subidentity: string;
  recipient_subidentity: string;
  inbox: string;
  encryption: EncryptionMethod;
}

export interface ExternalMetadata {
  sender: string;
  recipient: string;
  scheduled_time: string;
  signature: string;
  other: string;
}

export interface EncryptedShinkaiData {
  content: string;
}

export interface ShinkaiData {
  message_raw_content: string;
  message_content_schema: MessageSchemaType;
}

export type MessageData = { encrypted: EncryptedShinkaiData } | { unencrypted: ShinkaiData };

export interface EncryptedShinkaiBody {
  content: string;
}

export interface ShinkaiBody {
  message_data: MessageData;
  internal_metadata: InternalMetadata;
}

export type MessageBody = { encrypted: EncryptedShinkaiBody } | { unencrypted: ShinkaiBody };

export interface ShinkaiMessage {
  body: MessageBody | null;
  external_metadata: ExternalMetadata | null;
  encryption: EncryptionMethod;
}

export interface RegistrationCode {
  code: string;
  profileName: string;
  identityPk: string;
  encryptionPk: string;
  permissionType: string;
}
